File: typescript/day1/solution.ts
import fs from 'fs';
import path from 'path';

function readFile() {
  return fs.readFileSync(path.resolve(__dirname, './input.txt')).toString();
}

function getWeights(text: string) {
  const elves = text.split('\n\n');

  return elves.map((str) =>
    str
      .split('\n')
      .map((str) => parseInt(str))
      .reduce((sum, curr) => sum + curr)
  );
}

function solvePart1(text: string) {
  const weights = getWeights(text);

  return Math.max(...weights);
}

function solvePart2(text: string) {
  const weights = getWeights(text);

  const [one, two, three] = weights.sort((a, z) => z - a);

  return one + two + three;
}

const text = readFile();
console.log(`Part 1: ${solvePart1(text)}`);
console.log(`Part 2: ${solvePart2(text)}`);

File: golang/day1/solution.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func main() {
	dat, err := os.ReadFile("./day1/input.txt")
	check(err)

	txt := string(dat)

	elves := strings.Split(txt, "\n\n")
	totalElves := len(elves)
	totalWeights := make([]int, totalElves)
	for i := 0; i < totalElves; i++ {
		weightTexts := strings.Split(elves[i], "\n")
		sum := 0
		for j := 0; j < len(weightTexts); j++ {
			weight, err := strconv.Atoi(weightTexts[j])
			check(err)

			sum += weight
		}

		totalWeights[i] = sum
	}

	sort.Ints(totalWeights[:])

	fmt.Printf("Part 1: %d\n", totalWeights[totalElves-1])
	fmt.Printf("Part 2: %d\n", totalWeights[totalElves-1]+totalWeights[totalElves-2]+totalWeights[totalElves-3])
}
